### PR TITLE
Fix: FA2 with packed training

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes the issue from https://github.com/huggingface/transformers/pull/32241#issuecomment-2271314244. The FA2 was taking packed path when we were trying to continue with filled cache and had more than 1 new token. This could happen in assisted decoding for ex.

I think instead if checking the position ids length and last elements, we can check that if they are arranged in increasing order. We can know for sure that packed sequences will not have all elements with increasing positions. Also, we can be sure that in inference we'll always have an increasing order if there's no attn mask provided. If attn mask is there, we'll not reach this check so should not be a problem

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -264,11 +264,10 @@ def _flash_attention_forward(
...
 
-    # if position_ids is provided and check not all examples (row) contain only 1 sequence, and is in pre-fill/training stage
-    # then use `flash_attn_varlen_func` to prevent cross-example attention and also allow padding free approach
-    elif (
-        position_ids is not None and not (position_ids[:, -1] == position_ids.size(1) - 1).all() and query_length != 1
-    ):
+    # If position_ids is provided and check all examples do not contain only 1 sequence, If tensor in increasing
+    # then we probably have one sequence, otherwise it is packed. Additionally check we are in pre-fill/training stage.
+    # Use `flash_attn_varlen_func` to prevent cross-example attention and also allow padding free approach
+    elif position_ids is not None and not (torch.diff(position_ids) >= 0).all() and query_length != 1:
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:270]: can we explicit that we diff on the batch dim?

**Review by ArthurZucker — COMMENTED** [review]
We probably need a patch for this no? Can you add a non slow tests as well? 🤗

File: src/transformers/modeling_flash_attention_utils.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/modeling_flash_attention_utils.py:270]: yes, will do so

**Review by ArthurZucker — APPROVED** [review]
Cool! Can you just run the slow test to make sur it's not skipped! 🤗

**Comment by zucchini-nlp** [comment]
Slow CI tests are failing for unrelated reasons: gated repo, OOM or failing even w/o this PR. I believe this can be merged, since many tests need to be fixed in general

**Comment by ArthurZucker** [comment]
Yep we need to patch with this let's merge! 🤗

**Comment by zucchini-nlp** [comment]
Yep, and https://github.com/huggingface/transformers/pull/32527 for patch release plz. Can you approve it also?

**Pull request merged**